source/core/module_typeHelpers.js

```diff
@@ -0,0 +1,306 @@
+        /* eslint 'new-cap': ['error', {'capIsNewExceptions': [
+            'TypeRef_Name'
+        ]}], */
+        Module.typeHelpers = {};
+
+        var TypeRef_Name = Module.cwrap('TypeRef_Name', 'string', ['number']);
+
+        // Private instance functions
+        var validateVisitMethod = function (fn, fnName) {
+            if (typeof fn !== 'function') {
+                throw new Error('Visitor must have a method named `' + fnName + '`. Found: ' + fn);
+            }
+        };
+
+        var dispatchVisitBoolean = function (typeVisitor, valueRef, data) {
+            validateVisitMethod(typeVisitor.visitBoolean, 'visitBoolean');
+            return typeVisitor.visitBoolean(valueRef, data);
+        };
+
+        var dispatchVisitEnum = function (typeVisitor, valueRef, data) {
+            validateVisitMethod(typeVisitor.visitEnum, 'visitEnum');
+            return typeVisitor.visitEnum(typeVisitor, valueRef, data);
+        };
+
+        var dispatchVisitInteger = function (typeVisitor, valueRef, data) {
+            var typeRef = valueRef.typeRef;
+            var isSignedInteger = Module.typeHelpers.isSigned(typeRef);
+            var sizeOfInteger = Module.typeHelpers.topAQSize(typeRef);
+            var visitFn = undefined;
+            var typeName = '';
+            if (isSignedInteger === true) {
+                switch (sizeOfInteger) {
+                case 1:
+                    visitFn = typeVisitor.visitInt8;
+                    typeName = 'Int8';
+                    break;
+                case 2:
+                    visitFn = typeVisitor.visitInt16;
+                    typeName = 'Int16';
+                    break;
+                case 4:
+                    visitFn = typeVisitor.visitInt32;
+                    typeName = 'Int32';
+                    break;
+                case 8:
+                    visitFn = typeVisitor.visitInt64;
+                    typeName = 'Int64';
+                    break;
+                default:
+                    throw new Error('Unexpected size for Integer');
+                }
+            } else {
+                switch (sizeOfInteger) {
+                case 1:
+                    visitFn = typeVisitor.visitUInt8;
+                    typeName = 'UInt8';
+                    break;
+                case 2:
+                    visitFn = typeVisitor.visitUInt16;
+                    typeName = 'UInt16';
+                    break;
+                case 4:
+                    visitFn = typeVisitor.visitUInt32;
+                    typeName = 'UInt32';
+                    break;
+                case 8:
+                    visitFn = typeVisitor.visitUInt64;
+                    typeName = 'UInt64';
+                    break;
+                default:
+                    throw new Error('Unexpected size for Unsigned Integer. Found: ');
+                }
+            }
+
+            validateVisitMethod(visitFn, 'visit' + typeName);
```

Comment:
Can we move the visit concatenation too? The goal was to prevent the unnecessary allocation :P